ui-lib/src/main/java/cash/z/ecc/ui/MainActivity.kt

```diff
@@ -125,14 +147,45 @@ class MainActivity : ComponentActivity() {
     }
 
     @Composable
-    private fun WrapOnboarding() {
-        Onboarding(
-            onboardingState = onboardingViewModel.onboardingState,
-            onImportWallet = { TODO("Implement wallet import") },
-            onCreateWallet = {
-                walletViewModel.createAndPersistWallet()
+    private fun WrapRestore() {
+        val onboardingViewModel by viewModels<OnboardingViewModel>()
+        val restoreViewModel by viewModels<RestoreViewModel>()
+
+        when (val completeWordList = restoreViewModel.completeWordList.collectAsState().value) {
+            CompleteWordSetState.Loading -> {
+                // Although it might perform IO, it should be relatively fast.
+                // Consider whether to display indeterminate progress here.
+                // Another option would be to go straight to the restore screen with autocomplete
+                // disabled for a few milliseconds.  Users would probably never notice due to the
+                // time it takes to re-orient on the new screen, unless users were doing this
+                // on a daily basis and become very proficient at our UI.  The Therac-25 has
+                // historical precedent on how that could cause problems.
             }
-        )
+            is CompleteWordSetState.Loaded -> {
+                RestoreWallet(
+                    completeWordList.list,
+                    restoreViewModel.userWordList,
+                    onBack = { onboardingViewModel.isImporting.value = false },
+                    paste = {
+                        val clipboardManager = getSystemService(ClipboardManager::class.java)
+                        return@RestoreWallet clipboardManager?.primaryClip?.toString()
+                    },
+                    onFinished = {
+                        // Write the backup complete flag first, then the seed phrase.  That avoids the UI
+                        // flickering to the backup screen.  Assume if a user is restoring from
+                        // a backup, then the user has a valid backup.
+                        walletViewModel.persistBackupComplete()
+
+                        val restoredWallet = PersistableWallet(
+                            ZcashNetwork.fromResources(application),
+                            null,
```

Comment:
> This is where you should inject network's sapling activation height if no restore height is provided

Thanks for this good suggestion!

I'm thinking about where we should do this.  Implementing it here combines the logic to the wallet write, using a magic int as a sentinel value.  Later, we'd need `network.activationHeight == Wallet.birthday.height` to check it.

I wonder if there might be a few other options that could help us move away from the magic value in the stored representation?  Two ideas I had:
 - Store null and add an extension function like `fun WalletBirthday.birthdayOrFallback() = birthday ?: network.activationHeight`. This moves the determination to the wallet read, rather than the wallet write.  Callers can also check `WalletBirthday.birthday == null` to know that the real birthday is missing without checking the activation height.
 - Use a sealed class representation, with the birthday actually being either `class ActivationHeight()` or `class Birthday`.  This makes it a bit more object-oriented.

What do you think?

Reply:
The activation height is a protocol constant, so that number lives in the SDK and the wallet should use it as such :)

Reply:
I have made this change.

Eventually we may want to push this out of the UI, so that we can have stronger and more consistent representation invariants.  For example, the SDK already has partial logic for figuring out a fallback value when the birthday is null.  It would be great to complete this functionality in the SDK, and then defer to it.